src/main/java/org/broadinstitute/hellbender/engine/dataflow/transforms/PairReadWithRefBases.java

```diff
@@ -0,0 +1,51 @@
+package org.broadinstitute.hellbender.engine.dataflow.transforms;
+
+import com.google.cloud.dataflow.sdk.transforms.DoFn;
+import com.google.cloud.dataflow.sdk.transforms.PTransform;
+import com.google.cloud.dataflow.sdk.transforms.ParDo;
+import com.google.cloud.dataflow.sdk.values.KV;
+import com.google.cloud.dataflow.sdk.values.PCollection;
+import org.broadinstitute.hellbender.utils.SimpleInterval;
+import org.broadinstitute.hellbender.utils.read.GATKRead;
+import org.broadinstitute.hellbender.utils.reference.ReferenceBases;
+
+/**
+ * PairReadWithRefBases pairs the ReferenceBases overlapping the reads with each read and produces KV with read as
+ * the key.
+ *
+ * |---------- ref bases 1 ---------|      |-------------------- ref bases 2 --------------------|
+ * |-- read a --|   |---- read b ---|         |--------------- read c --------------|
+ * results in,
+ *
+ * |-- read a --|   |---- read b ---|         |--------------- read c --------------|
+ * |ref bases 1a|   |-ref bases 1b -|         |----------- ref bases 2c ------------|
+ *
+ * KV<read a, ref bases 1a>
+ * KV<read b, ref bases 1b>
+ * KV<read c, ref bases 2c>
+ * Precondition: each refbases must contain ALL of the bases for all of the GATKReads it's paired with.
+ */
```

Comment:
@jean-philippe-martin, I don't quite follow your first comment.
If a read spans to ref shards, it's mapped to the first and it's span impact which ref bases are queries for. It has no impact on the second shard. Let me know how I should update the comment.